File: app/src/main/java/indie/mefistofel/game15/GameActivity.java
package indie.mefistofel.game15;

import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.support.design.widget.NavigationView;
import android.support.v4.view.GravityCompat;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;

public class GameActivity extends AppCompatActivity
        implements NavigationView.OnNavigationItemSelectedListener, IFragmentInteractionListener {

    private Fragment mCurrentFragment;
    private static final String TAG_CURRENT_FRAGMENT = "current_fragment";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_game);
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        DrawerLayout drawer = (DrawerLayout) findViewById(R.id.drawer_layout);
        ActionBarDrawerToggle toggle = new ActionBarDrawerToggle(
                this, drawer, toolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close);
        drawer.addDrawerListener(toggle);
        toggle.syncState();

        NavigationView navigationView = (NavigationView) findViewById(R.id.nav_view);
        navigationView.setNavigationItemSelectedListener(this);

        FragmentManager fragmentManager = getSupportFragmentManager();
        mCurrentFragment = fragmentManager.findFragmentByTag(TAG_CURRENT_FRAGMENT);
        if (mCurrentFragment == null) {
            shuffleNewGame();
        }
    }

    @Override
    public void onBackPressed() {
        DrawerLayout drawer = (DrawerLayout) findViewById(R.id.drawer_layout);
        if (drawer.isDrawerOpen(GravityCompat.START)) {
            drawer.closeDrawer(GravityCompat.START);
        } else {
            super.onBackPressed();
        }
    }

    @Override
    public boolean onNavigationItemSelected(@NonNull MenuItem item) {
        int id = item.getItemId();
        if (id == R.id.nav_shuffle) {
            shuffleNewGame();
        } else if (id == R.id.nav_info) {
            onFragmentInteraction(Action.info);
        }

        DrawerLayout drawer = (DrawerLayout) findViewById(R.id.drawer_layout);
        drawer.closeDrawer(GravityCompat.START);
        return true;
    }

    @Override
    public void onFragmentInteraction(Action action) {
        switch (action) {
            case win:
                showFragment(new WinFragment());
                break;
            case shuffle:
                shuffleNewGame();
                break;
            case info:
                showFragment(new InfoFragment(), true);
                break;
        }
    }

    private void shuffleNewGame() {
        FieldData.getInstance().shuffleFields();
        showFragment(new GameFragment());
    }

    void showFragment(Fragment fragment) {
        showFragment(fragment, false);
    }

    void showFragment(Fragment fragment, boolean addToBackStack) {
        mCurrentFragment = fragment;
        FragmentManager myFragmentManager = getSupportFragmentManager();
        FragmentTransaction fragmentTransaction = myFragmentManager.beginTransaction();
        fragmentTransaction.replace(R.id.container, fragment, TAG_CURRENT_FRAGMENT);
        fragmentTransaction.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_OPEN);
        if (addToBackStack) {
            fragmentTransaction.addToBackStack(null);
        }
        fragmentTransaction.commit();
    }
}

File: Game15Project/app/src/main/java/indie/mefistofel/game15/GameActivity.kt
package indie.mefistofel.game15

import android.os.Bundle
import android.view.KeyEvent
import android.view.MenuItem
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.core.view.GravityCompat
import androidx.drawerlayout.widget.DrawerLayout
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentTransaction
import com.google.android.material.navigation.NavigationView
import indie.mefistofel.game15.IFragmentInteractionListener.*

class GameActivity : AppCompatActivity(), NavigationView.OnNavigationItemSelectedListener,
    IFragmentInteractionListener {
    private var mCurrentFragment: Fragment? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_game)
        val toolbar: Toolbar = findViewById(R.id.toolbar)
        setSupportActionBar(toolbar)
        val drawer: DrawerLayout = findViewById(R.id.drawer_layout)
        val toggle = ActionBarDrawerToggle(
            this, drawer, toolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close
        )
        drawer.addDrawerListener(toggle)
        toggle.syncState()
        val navigationView: NavigationView = this.findViewById(R.id.nav_view)
        navigationView.setNavigationItemSelectedListener(this)
        val fragmentManager: FragmentManager = supportFragmentManager
        mCurrentFragment = fragmentManager.findFragmentByTag(TAG_CURRENT_FRAGMENT)
        if (mCurrentFragment == null) {
            shuffleNewGame()
        }
    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent?): Boolean {
        if (mCurrentFragment is IFragmentKey) {
            (mCurrentFragment as IFragmentKey?)!!.OnKeyDown(keyCode)
        }
        return super.onKeyDown(keyCode, event)
    }

    override fun onBackPressed() {
        val drawer: DrawerLayout = findViewById(R.id.drawer_layout)
        if (drawer.isDrawerOpen(GravityCompat.START)) {
            drawer.closeDrawer(GravityCompat.START)
        } else {
            super.onBackPressed()
        }
    }

    override fun onNavigationItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        if (id == R.id.nav_shuffle) {
            shuffleNewGame()
        } else if (id == R.id.nav_info) {
            onFragmentInteraction(Action.info)
        }
        val drawer: DrawerLayout = findViewById(R.id.drawer_layout)
        drawer.closeDrawer(GravityCompat.START)
        return true
    }

    override fun onFragmentInteraction(action: Action) {
        when (action) {
            Action.win -> showFragment(WinFragment())
            Action.shuffle -> shuffleNewGame()
            Action.info -> showFragment(InfoFragment(), true)
        }
    }

    private fun shuffleNewGame() {
        FieldData.getInstance().shuffleFields()
        showFragment(GameFragment())
    }

    fun showFragment(fragment: Fragment) {
        showFragment(fragment, false)
    }

    fun showFragment(fragment: Fragment, addToBackStack: Boolean) {
        mCurrentFragment = fragment
        val myFragmentManager: FragmentManager = getSupportFragmentManager()
        val fragmentTransaction: FragmentTransaction = myFragmentManager.beginTransaction()
        fragmentTransaction.replace(R.id.container, fragment, TAG_CURRENT_FRAGMENT)
        fragmentTransaction.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_OPEN)
        if (addToBackStack) {
            fragmentTransaction.addToBackStack(null)
        }
        fragmentTransaction.commit()
    }

    companion object {
        private const val TAG_CURRENT_FRAGMENT = "current_fragment"
    }
}